src/ip_notifier.cpp

```diff
@@ -360,40 +359,52 @@ struct ip_change_notifier_impl final : ip_change_notifier
 
 	~ip_change_notifier_impl() override
 	{
-		cancel();
-		// the reason to call close() here is because the
-		// object_handle destructor doesn't close the handle
-		m_hnd.close();
+		if (m_hnd != NULL)
+		{
+			CancelMibChangeNotify2(m_hnd);
+			m_hnd = NULL;
+		}
 	}
 
 	void async_wait(std::function<void(error_code const&)> cb) override
 	{
-		HANDLE hnd;
-		DWORD err = NotifyAddrChange(&hnd, &m_ovl);
-		if (err == ERROR_IO_PENDING)
-		{
-			m_hnd.async_wait([cb](error_code const& ec)
-			{
-				// call CancelIPChangeNotify here?
-				cb(ec);
-			});
-		}
-		else
+		if (m_hnd == NULL)
 		{
-			lt::get_io_service(m_hnd).post([cb, err]()
-			{ cb(error_code(err, system_category())); });
+			cb(make_error_code(boost::system::errc::not_supported));
+			return;
 		}
+
+		std::lock_guard<std::mutex> l(m_cb_mutex);
+		m_cb.emplace_back(std::move(cb));
 	}
 
 	void cancel() override
 	{
-		CancelIPChangeNotify(&m_ovl);
-		m_hnd.cancel();
+		std::vector<std::function<void(error_code const&)>> cbs;
+		{
+			std::lock_guard<std::mutex> l(m_cb_mutex);
+			cbs = std::move(m_cb);
```

Comment:
I forget, is `std::vector` guaranteed to be empty after move? I think it needs to be here, right?

Reply:
Yes, moved-from vectors are always empty.